examples/stm32/f4/stm32f429i-discovery/lcd-ltdc/gfx_locm3.c

```diff
@@ -0,0 +1,645 @@
+	__gfx_state.wrap         = true;
+	__gfx_state.surface      = surface;
+	__gfx_state.font         = NULL;
+}
+void gfx_set_surface(uint16_t *surface) {
+	__gfx_state.surface = surface;
+}
+uint16_t *
+gfx_get_surface() {
+	return __gfx_state.surface;
+}
+
+void
+gfx_set_surface_visible_area_max() {
+	__gfx_state.visible_area = (visible_area_t){0,0,__gfx_state.width,__gfx_state.height};
+}
+void
+gfx_set_surface_visible_area(int16_t x1, int16_t y1, int16_t x2, int16_t y2) {
+	if (x1<0)  x1=0;
+	if (y1<0)  y1=0;
+	if (x2<x1) x2=x1;
+	if (y2<y1) y2=y1;
+	__gfx_state.visible_area = (visible_area_t){x1,y1,x2,y2};
+}
+visible_area_t
+gfx_get_surface_visible_area() {
+	return __gfx_state.visible_area;
+}
+
+
+/*
+ * just write a RGB565 color to each pixel in the current buffer/surface
+ */
+void gfx_fill_screen(uint16_t color) {
+	uint32_t i; uint16_t *pixel_addr;
+	pixel_addr = __gfx_state.surface;
+	for (i = 0; i < __gfx_state.pixel_count; i++) {
+		*pixel_addr++ = color;
+	}
+}
+
+
+
+/* change the rotation and flip width and height accordingly */
+void gfx_rotate(gfx_rotation_t rotation) {
+	if (rotation == GFX_ROTATION_0_DEGREES || rotation == GFX_ROTATION_180_DEGREES) {
+		__gfx_state.width  = __gfx_state.width_orig;
+		__gfx_state.height = __gfx_state.height_orig;
+	} else {
+		__gfx_state.width  = __gfx_state.height_orig;
+		__gfx_state.height = __gfx_state.width_orig;
+	}
+	__gfx_state.rotation = rotation;
+	gfx_set_surface_visible_area_max();
+}
+uint8_t gfx_get_rotation(void) {
+  return __gfx_state.rotation;
+}
+// Return the size of the display (per current rotation)
+uint16_t gfx_width(void) {
+  return __gfx_state.width;
+}
+
+uint16_t gfx_height(void) {
+  return __gfx_state.height;
+}
+
+
+/*
+ * draw a single pixel
+ * changes buffer addressing (surface) according to orientation
+ */
+void gfx_draw_pixel(int16_t x, int16_t y, uint16_t color) {
+	uint16_t *pixel_addr;
+	if (x < __gfx_state.visible_area.x1 || x >= __gfx_state.visible_area.x2) {
+		return;
+	}
+	if (y < __gfx_state.visible_area.y1 || y >= __gfx_state.visible_area.y2) {
+		return;
+	}
+
+	pixel_addr = __gfx_state.surface;
+	switch (__gfx_state.rotation) {
+		case GFX_ROTATION_0_DEGREES :
+			pixel_addr +=                           (x + y*__gfx_state.width_orig);
+			break;
+		case GFX_ROTATION_180_DEGREES :
+			pixel_addr += __gfx_state.pixel_count - (x + y*__gfx_state.width_orig);
+			break;
+		case GFX_ROTATION_90_DEGREES :
+			pixel_addr += __gfx_state.pixel_count - (x*__gfx_state.width_orig + (__gfx_state.width_orig-y));
+			break;
+		case GFX_ROTATION_270_DEGREES :
+			pixel_addr +=                           (x*__gfx_state.width_orig + (__gfx_state.width_orig-y));
+			break;
+	}
+	*pixel_addr = color;
+
+	return;
+}
+
+// Bresenham's algorithm - thx wikpedia
+void gfx_draw_line(int16_t x0, int16_t y0,
+			    int16_t x1, int16_t y1,
+			    uint16_t fg) {
+  int16_t steep = abs(y1 - y0) > abs(x1 - x0);
+  if (steep) {
+    swap(x0, y0);
+    swap(x1, y1);
+  }
+

```

Comment:
Ollie, when you look at the Adafruit Graphics library fro draw line it has:
```

// Bresenham's algorithm - thx wikpedia
void Adafruit_GFX::drawLine(int16_t x0, int16_t y0, int16_t x1, int16_t y1,
 uint16_t color) {
  int16_t steep = abs(y1 - y0) > abs(x1 - x0);
  if (steep) {
    _swap_int16_t(x0, y0);
    _swap_int16_t(x1, y1);
}
...
```
I suggest to avoid misunderstandings (and possibly litigation) in the future that you might want to express thanks to the Adafruit folks for code that helped you write this (Theirs is a pretty handy little library class, I started with it myself when I was putting together the simple graphics library repo).